src/Nethermind/Nethermind.Core/Caching/ConcurrentLruCache.cs

```diff
@@ -0,0 +1,266 @@
+    {
+        private readonly int _maxCapacity;
+        private readonly Dictionary<TKey, LinkedListNode<LruCacheItem>> _cacheMap;
+        private readonly ReadWriteLockDisposable _lock = new();
+        private readonly ConcurrentQueue<LinkedListNode<LruCacheItem>> _accesses;
+
+        private LinkedListNode<LruCacheItem>? _leastRecentlyUsed;
+        private CancellationTokenSource _cts;
+
+        private long _toProcess;
+        private int _processingLru;
+
+        public ConcurrentLruCache(int maxCapacity, int startCapacity, string name)
+        {
+            // Should use NonBlockingLruCache if key is less than or equal to pointer size.
+            ArgumentOutOfRangeException.ThrowIfLessThanOrEqual(Unsafe.SizeOf<TKey>(), IntPtr.Size);
+            ArgumentOutOfRangeException.ThrowIfLessThan(maxCapacity, 1);
+
+            _maxCapacity = maxCapacity;
+            _accesses = new ConcurrentQueue<LinkedListNode<LruCacheItem>>();
+            _cacheMap = typeof(TKey) == typeof(byte[])
+                ? new Dictionary<TKey, LinkedListNode<LruCacheItem>>((IEqualityComparer<TKey>)Bytes.EqualityComparer)
+                : new Dictionary<TKey, LinkedListNode<LruCacheItem>>();
+            _cts = new CancellationTokenSource();
+        }
+
+        public ConcurrentLruCache(int maxCapacity, string name)
+            : this(maxCapacity, 0, name)
+        {
+        }
+
+        public void Clear()
+        {
+            // Signal background thread to stop
+            _cts.Cancel();
+            // Signal background thread to stop (two ways)
+            _leastRecentlyUsed = null;
+
+            using var handle = _lock.AcquireWrite();
+
+            _accesses.Clear();
+            _cacheMap.Clear();
+            _cts = new CancellationTokenSource();
+            // reclear incase background thread set it before accesses were cleared
+            _leastRecentlyUsed = null;
+        }
+
+        public TValue Get(TKey key) => TryGet(key, out TValue value) ? value : default!;
+
+        public bool TryGet(TKey key, out TValue value)
+        {
+            bool didGet;
+            LinkedListNode<LruCacheItem>? node;
+            using (var handle = _lock.AcquireWrite())
+            {
+                didGet = _cacheMap.TryGetValue(key, out node);
+            }
+
+            if (didGet)
+            {
+                value = node!.Value.Value;
+                ScheduleLruAccounting(node);
+                return true;
+            }
+
+            value = default!;
+            return false;
+        }
+
+        public bool Set(TKey key, TValue val)
+        {
+            if (val is null)
+            {
+                Delete(key);
+            }
+
+            ref LinkedListNode<LruCacheItem>? node = ref Unsafe.NullRef<LinkedListNode<LruCacheItem>?>();
+            using (var handle = _lock.AcquireWrite())
+            {
+                node = ref CollectionsMarshal.GetValueRefOrAddDefault(_cacheMap, key, out _);
```

Comment:
This is a ref to an array. If another thread resizes the array, this ref will point to an array no longer in use. Am I missing something or is it a bug?